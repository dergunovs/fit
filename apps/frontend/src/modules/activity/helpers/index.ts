import { IActivity, IExerciseDone, IExerciseStatistics } from 'fitness-tracker-contracts';
import { createTempId, formatDate, formatDuration, subtractDates } from 'mhz-helpers';
import { toast } from 'mhz-ui';

export function getPotentialActivityDuration(
  exercises: IExerciseDone[],
  exerciseStatistics: IExerciseStatistics[],
  averageRestPercent?: number
) {
  if (!averageRestPercent) return '-';

  const totalDuration = exercises.reduce((acc, exercise) => {
    const averageDuration =
      exerciseStatistics.find((choosenExericse) => choosenExericse._id === exercise.exercise?._id)?.averageDuration ||
      0;

    return acc + averageDuration * exercise.repeats;
  }, 0);

  const durationWithRest = Math.round(totalDuration / (1 - averageRestPercent / 100));

  return formatDuration(durationWithRest);
}

export function convertActivityCalendarEvents(activities?: IActivity[]) {
  return activities?.map((activity: IActivity) => {
    return {
      _id: activity._id,
      start: new Date(`${activity.dateCreated}`),
      end: new Date(`${activity.dateUpdated}`),
      title: '+',
      content: activity.exercises,
    };
  });
}

export function generateActivityExercises(exercisesDone: IExerciseDone[]): IExerciseDone[] {
  const activityExercises = exercisesDone.map((exercise) => {
    return {
      _id: createTempId(),
      exercise: {
        _id: exercise.exercise?._id || '',
        title: exercise.exercise?.title || '',
        muscleGroups: exercise.exercise?.muscleGroups || [],
        isWeights: exercise.exercise?.isWeights || false,
        isWeightsRequired: exercise.exercise?.isWeightsRequired || false,
      },
      repeats: exercise.repeats,
      weight: exercise.weight,
    };
  });

  return activityExercises?.length ? [...activityExercises] : [];
}

export function getToFailurePercent(exercises: IExerciseDone[]) {
  const allExercises = exercises.length;
  const toFailureExercises = exercises.filter((exercise) => exercise.isToFailure).length;

  return `${Math.floor((toFailureExercises / allExercises) * 100)}%`;
}

export function getRestPercent(exercises: IExerciseDone[], start?: Date | null | string, end?: Date | null | string) {
  const activityDuration = Number(subtractDates(end, start, true));
  const exercisesDuration = exercises.reduce((acc, current) => acc + (current.duration || 0), 0);

  return `${Math.floor(100 - (exercisesDuration / activityDuration) * 100)}%`;
}

export async function copyActivityToClipboard(
  exercises: IExerciseDone[],
  start?: Date | null | string,
  end?: Date | null | string
) {
  const textHeader = `${formatDate(start, 'ru')}, длительность: ${subtractDates(end, start)}
Подходы: ${exercises.length}, отказы: ${getToFailurePercent(exercises)}, отдых: ${getRestPercent(exercises, start, end)}.

${exercises
  .map((exercise, index) => {
    return `${index + 1}. ${exercise.exercise?.title} x${exercise.repeats} ${exercise.weight ? `${exercise.weight}кг` : ''} ${formatDuration(exercise.duration)} ${exercise.isToFailure ? 'ДО ОТКАЗА' : ''}\n`;
  })
  .join('')}`;

  await navigator.clipboard.writeText(textHeader);

  toast.success('Скопировано в буфер');
}
