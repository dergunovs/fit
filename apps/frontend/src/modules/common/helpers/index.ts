import { toast } from 'mhz-ui';

export async function copyToClipboard(text: string, toastText: string) {
  await navigator.clipboard.writeText(text);

  toast.success(toastText);
}
