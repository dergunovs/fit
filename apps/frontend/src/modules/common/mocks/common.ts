import { vi } from 'vitest';

export const spyCopyToClipboard = vi.spyOn(navigator.clipboard, 'writeText').mockResolvedValue();
