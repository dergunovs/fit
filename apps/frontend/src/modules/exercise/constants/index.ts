export const URL_EXERCISE = '/exercises';
export const URL_EXERCISE_CREATE = `${URL_EXERCISE}/create`;
export const URL_EXERCISE_EDIT = `${URL_EXERCISE}/edit`;

export const EXERCISE_REPEATS_OPTIONS = [8, 10, 12, 15, 20];
export const EXERCISE_REPEATS_DEFAULT = 12;

export const EXERCISE_STATISTICS_HEADERS = [{ title: 'Упражнение' }, { title: 'Подходы' }, { title: 'Повторы' }];
export const MUSCLE_GROUP_STATISTICS_HEADERS = [{ title: 'Мышцы' }, { title: 'Подходы' }, { title: 'Повторы' }];
export const EXERCISE_LIST_HEADERS = [{ title: 'Название' }];

export const EXERCISE_DELETED_TITLE = 'Упражнение удалено';
