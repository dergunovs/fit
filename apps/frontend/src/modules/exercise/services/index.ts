import { ComputedRef, Ref } from 'vue';
import {
  API_EXERCISE,
  API_EXERCISE_ALL,
  API_EXERCISE_CUSTOM,
  TGetExercisesDTO,
  TGetExercisesCustomDTO,
  TGetExercisesQueryDTO,
  TGetExerciseDTO,
  TPostExerciseDTO,
  TPostExerciseDataDTO,
  TUpdateExerciseDTO,
  TUpdateExerciseDataDTO,
  TDeleteExerciseDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const exerciseService = {
  getMany: (page: Ref<number>) => {
    return useQuery({
      queryKey: [API_EXERCISE, page],
      queryFn: async () => {
        const params: TGetExercisesQueryDTO = { page: page.value };

        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE, { params });

        return data;
      },
    });
  },

  getAll: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_ALL],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE_ALL);

        return data.data;
      },
    });
  },

  getCustom: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_CUSTOM],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesCustomDTO>(API_EXERCISE_CUSTOM);

        return data.data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined>) => {
    return useQuery({
      queryKey: [API_EXERCISE, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetExerciseDTO>(`${API_EXERCISE}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TPostExerciseDataDTO) => {
        const { data } = await api.post<TPostExerciseDTO>(API_EXERCISE, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TUpdateExerciseDataDTO) => {
        const { data } = await api.patch<TUpdateExerciseDTO>(`${API_EXERCISE}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteExerciseDTO>(`${API_EXERCISE}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
