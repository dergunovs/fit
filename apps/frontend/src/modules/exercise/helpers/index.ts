import {
  IExercise,
  IExerciseChoosen,
  IExerciseDone,
  IExerciseStatistics,
  IUser,
  IUserEquipment,
} from 'fitness-tracker-contracts';
import { createTempId } from 'mhz-helpers';

import { ITimelineStep } from '@/activity/interface';
import { getWeightsForUserEquipment } from '@/equipment/helpers';

export function getUserEquipmentParams(exercise: IExercise, user?: IUser) {
  const isExerciseHasEquipment = !!exercise.equipment;
  const isExerciseHasEquipmentForWeight = !!exercise.equipmentForWeight?.length;
  const isWeightsRequired = !!exercise.isWeightsRequired;

  const isUserHasEquipment = !!user?.equipments?.some(
    (equipment) => equipment.equipment?._id?.toString() === exercise.equipment?._id?.toString()
  );

  const isUserHasEquipmentForWeight = !!user?.equipments?.some((equipment) =>
    exercise.equipmentForWeight?.some(
      (equipmentForWeight) => equipmentForWeight._id?.toString() === equipment.equipment?._id?.toString()
    )
  );

  return {
    isExerciseHasEquipment,
    isExerciseHasEquipmentForWeight,
    isWeightsRequired,
    isUserHasEquipment,
    isUserHasEquipmentForWeight,
  };
}

export function isUserEquipmentMatches(exercise: IExercise, user?: IUser) {
  let result = false;

  const {
    isExerciseHasEquipment,
    isExerciseHasEquipmentForWeight,
    isWeightsRequired,
    isUserHasEquipment,
    isUserHasEquipmentForWeight,
  } = getUserEquipmentParams(exercise, user);

  if (!isExerciseHasEquipment && !isExerciseHasEquipmentForWeight) {
    result = true;
  } else if (isUserHasEquipment && !isWeightsRequired) {
    result = true;
  } else if (!isExerciseHasEquipment && isExerciseHasEquipmentForWeight && isUserHasEquipmentForWeight) {
    result = true;
  } else if (!isExerciseHasEquipment && !isWeightsRequired) {
    result = true;
  } else if (
    isExerciseHasEquipment &&
    isExerciseHasEquipmentForWeight &&
    isUserHasEquipment &&
    isUserHasEquipmentForWeight
  ) {
    result = true;
  }

  return result;
}

export function getAverageDuration(exercise: IExerciseStatistics, type: 'set' | 'repeat') {
  return type === 'set'
    ? `${((exercise.averageDuration * exercise.repeats) / exercise.sets || 0).toFixed(1)}`
    : `${exercise.averageDuration.toFixed(1)}`;
}

export function getAverageRepeatsInSet(repeats?: number, sets?: number) {
  return repeats && sets ? Math.round(repeats / sets) : 0;
}

export function isPrevExerciseSame(exercises: IExerciseDone[], index: number, id?: string) {
  return id && exercises[index - 1] ? id === exercises[index - 1].exercise?._id : false;
}

export function isSetCreatable(choosenExercises: IExerciseChoosen[], index: number, id?: string) {
  return !isPrevExerciseSame(choosenExercises, index, id) && index > 0 && index + 1 === choosenExercises.length;
}

export function generateTimeline(exercises: IExerciseDone[], start: Date | string | null, ratio: number) {
  const allSteps: ITimelineStep[] = [];

  exercises.forEach((exercise, index) => {
    if (!exercise.dateUpdated || !exercise.duration || !start) return;

    allSteps.push({
      left:
        index === 0
          ? 0
          : (new Date(exercise.dateUpdated).getTime() - exercise.duration * 1000 - new Date(start).getTime()) / ratio,
      right:
        index === exercises.length - 1 && exercises.length !== 1
          ? 0
          : (new Date(exercise.dateUpdated).getTime() - new Date(start).getTime()) / ratio,
      type: 'exercise',
    });
  });

  allSteps.forEach((step, index) => {
    if (!allSteps[index - 1]?.right) return;

    allSteps.push({ left: allSteps[index - 1].right, right: step.left, type: 'rest' });
  });

  return allSteps.sort((a, b) => a.left - b.left);
}

export function filterExercisesByTitleAndMuscle(exercises: IExercise[], title: string, muscle?: string, user?: IUser) {
  return exercises.filter((exercise) => {
    const isEquipmentMatches = isUserEquipmentMatches(exercise, user);

    const titleFilter =
      (exercise.title.toLowerCase().includes(title.toLocaleLowerCase()) ||
        exercise.title_en?.toLowerCase().includes(title.toLocaleLowerCase())) &&
      isEquipmentMatches;
    const muscleFilter = exercise.muscles?.some((group) => group._id === muscle) && isEquipmentMatches;

    return muscle ? muscleFilter && titleFilter : titleFilter;
  });
}

export function getAvailableExerciseWeights(exercise: IExercise, user: IUser) {
  const equipments = user.equipments?.filter((equipment) =>
    exercise.equipmentForWeight?.some((eq) => eq._id === equipment.equipment?._id)
  );

  if (!equipments?.length) return [];

  const weights = getWeightsForUserEquipment(equipments);

  return exercise.isWeightsRequired ? weights : [0, ...weights];
}

export function getExercisesToChooseDefaultWeight(exercises: IExercise[], userEquipments?: IUserEquipment[]) {
  const equipmentsForWeight = userEquipments?.filter((eq) => eq.weights?.length);
  const exercisesWithWeight = exercises.filter((ex) => ex.isWeights);

  const tableData = exercisesWithWeight.map((ex) => {
    const exerciseEquipment = ex.equipmentForWeight
      ?.filter((eq) => equipmentsForWeight?.some((equipment) => equipment.equipment?._id === eq._id))
      .map((exToMap) => exToMap._id);

    const availableEquipment = equipmentsForWeight?.filter((eq) =>
      exerciseEquipment?.some((equipmentId) => equipmentId === eq.equipment?._id)
    );

    const options = getWeightsForUserEquipment(availableEquipment);

    return {
      _id: ex._id,
      title: ex.title,
      title_en: ex.title_en,
      options: ex.isWeightsRequired ? options : [0, ...options],
    };
  });

  return tableData.filter((ex) => ex.options.length > 1);
}

export function getDefaultExerciseWeight(exercise: IExercise, user: IUser, weights?: number[]) {
  let weight = weights?.length && exercise.isWeightsRequired ? weights[0] : 0;

  if (user.defaultWeights && exercise._id && Object.hasOwn(user.defaultWeights, exercise._id)) {
    weight = user.defaultWeights[exercise._id];
  }

  return weight;
}

export function addSetToExercises(exercises: IExerciseChoosen[]) {
  const set = exercises.slice(-2).map((exercise) => {
    return { ...exercise, _id: createTempId() };
  });

  return [...exercises, ...set];
}

export function updateExercisesIndex(exercises: IExerciseChoosen[], index: number) {
  if (index <= 0 || index >= exercises.length) return exercises;

  const updatedExercises = [...exercises];
  const exerciseToMove = updatedExercises[index];

  updatedExercises.splice(index, 1);
  updatedExercises.splice(index - 1, 0, exerciseToMove);

  return updatedExercises;
}

export function updateExerciseField<T extends keyof IExerciseDone>(
  exercises: IExerciseChoosen[],
  field: T,
  value: IExerciseDone[T],
  id?: string
): IExerciseChoosen[] {
  return exercises.map((exercise) => (exercise._id === id ? { ...exercise, [field]: value } : exercise));
}

export function generateExerciseChoosen(repeats: number, weight?: number, exercise?: IExercise): IExerciseChoosen {
  return {
    _id: createTempId(),
    exercise: {
      _id: exercise?._id || '',
      title: exercise?.title || '',
      title_en: exercise?.title_en || '',
      muscles: exercise?.muscles || [],
      isWeights: exercise?.isWeights || false,
      isWeightsRequired: exercise?.isWeightsRequired || false,
      equipmentForWeight: exercise?.equipmentForWeight,
    },
    repeats,
    weight,
  };
}
