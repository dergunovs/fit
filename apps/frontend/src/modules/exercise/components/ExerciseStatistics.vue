<template>
  <div>
    <UiTable :headers="EXERCISE_STATISTICS_HEADERS" :lang="locale">
      <tr v-for="exercise in props.statistics" :key="exercise.exercise._id" data-test="exercise-statistics-table-row">
        <td>
          <UiButton
            @click="showExercise(exercise)"
            layout="plain"
            isWrap
            data-test="exercise-statistics-title"
            :class="$style.title"
            :data-equipment="isAuth ? exercise.isUserEquipmentMatches : true"
            :data-custom="exercise.exercise.isCustom"
          >
            {{ exercise.exercise[localeField('title', locale)] }}
          </UiButton>
        </td>
        <td>
          <div :class="$style.cell">
            <div :class="$style.count" data-test="exercise-statistics-sets-count">
              {{ exercise.sets }}<DynamicPercent :percent="exercise.setsDynamics" />
            </div>
            <div :class="$style.duration" data-test="exercise-statistics-sets-duration">
              {{ getAverageDuration(exercise, 'set') }}{{ t('sec') }}
            </div>
          </div>
        </td>
        <td>
          <div :class="$style.cell">
            <div :class="$style.count" data-test="exercise-statistics-repeats-count">
              {{ exercise.repeats }}<DynamicPercent :percent="exercise.repeatsDynamics" />
            </div>
            <div :class="$style.duration" data-test="exercise-statistics-repeats-in-set">
              {{ getAverageRepeatsInSet(exercise.repeats, exercise.sets) }}
            </div>
          </div>
        </td>
      </tr>
    </UiTable>

    <UiModal v-model="isShowModal" data-test="exercise-statistics-modal">
      <ExerciseInfo v-if="currentExercise" :exercise="currentExercise" data-test="exercise-statistics-info" />
    </UiModal>
  </div>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';
import { IExerciseStatistics } from 'fitness-tracker-contracts';
import { UiTable, UiButton, UiModal } from 'mhz-ui';
import { isAuth, localeField } from 'mhz-helpers';

import ExerciseInfo from '@/exercise/components/ExerciseInfo.vue';
import DynamicPercent from '@/common/components/DynamicPercent.vue';

import { getAverageDuration, getAverageRepeatsInSet } from '@/exercise/helpers';
import { useTI18n } from '@/common/composables';

interface IProps {
  statistics: IExerciseStatistics[];
}

const props = defineProps<IProps>();

const { t, locale } = useTI18n();

const isShowModal = ref(false);

const currentExercise = ref<IExerciseStatistics>();

const EXERCISE_STATISTICS_HEADERS = computed(() => [
  { title: t('exercise.one') },
  { title: t('set.many') },
  { title: t('repeat.many') },
]);

function showExercise(exercise: IExerciseStatistics) {
  currentExercise.value = exercise;

  isShowModal.value = true;
}
</script>

<style module lang="scss">
.cell {
  display: flex;
  gap: 4px;
}

.title {
  &[data-equipment='false'] {
    color: var(--color-gray-dark);
  }

  &[data-custom='true'] {
    font-weight: 700;
  }
}

.count {
  display: flex;
  flex-wrap: nowrap;
}

.duration {
  align-self: flex-end;
  font-size: 0.875rem;
  color: var(--color-gray-dark);
}

@media (max-width: 960px) {
  .cell {
    flex-direction: column;
  }

  .duration {
    align-self: flex-start;
  }
}

:global(.dark) {
  .count {
    color: var(--color-white);
  }

  .title {
    &[data-equipment='false'] {
      color: var(--color-gray-dark);
    }
  }

  .duration {
    color: var(--color-gray);
  }
}
</style>
