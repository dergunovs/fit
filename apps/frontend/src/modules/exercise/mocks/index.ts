import { vi } from 'vitest';
import {
  TDeleteExerciseDTO,
  TGetExercisesDTO,
  TPostExerciseDataDTO,
  TPostExerciseDTO,
  TUpdateExerciseDataDTO,
  TUpdateExerciseDTO,
} from 'fitness-tracker-contracts';

import { exerciseService } from '@/exercise/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { EXERCISE_FIXTURE, EXERCISE_FIXTURE_CUSTOM, EXERCISES_FIXTURE } from '@/exercise/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetExercise = vi.spyOn(exerciseService, 'getOne').mockReturnValue(mockQueryReply(EXERCISE_FIXTURE));

const mockGetExercisesData: TGetExercisesDTO = { data: EXERCISES_FIXTURE, total: EXERCISES_FIXTURE.length };

const spyGetExercises = vi
  .spyOn(exerciseService, 'getMany')
  .mockImplementation(() => mockQueryReply(mockGetExercisesData));

const spyGetExercisesAll = vi
  .spyOn(exerciseService, 'getAll')
  .mockImplementation(() => mockQueryReply(EXERCISES_FIXTURE));

const spyGetExercisesCustom = vi
  .spyOn(exerciseService, 'getCustom')
  .mockImplementation(() => mockQueryReply([EXERCISE_FIXTURE_CUSTOM]));

const spyCreateExercise = vi.fn();
const spyUpdateExercise = vi.fn();
const spyDeleteExercise = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(exerciseService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostExerciseDTO, TPostExerciseDataDTO>(spyCreateExercise);
});

vi.spyOn(exerciseService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateExerciseDTO, TUpdateExerciseDataDTO>(spyUpdateExercise);
});

vi.spyOn(exerciseService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteExerciseDTO, string>(spyDeleteExercise);
});

export {
  spyGetExercise,
  spyGetExercises,
  spyGetExercisesCustom,
  spyGetExercisesAll,
  spyCreateExercise,
  spyUpdateExercise,
  spyDeleteExercise,
  mockOnSuccess,
  mockGetExercisesData,
};
