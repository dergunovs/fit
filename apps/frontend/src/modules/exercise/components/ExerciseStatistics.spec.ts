import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';

import ExerciseStatistics from './ExerciseStatistics.vue';

import { dataTest, wrapperFactory } from '@/common/test';
import { EXERCISES_STATISTICS_FIXTURE } from '@/exercise/fixtures';
import { getAverageDuration } from '@/exercise/helpers';

const tableRow = dataTest('exercise-statistics-table-row');
const title = dataTest('exercise-statistics-title');
const setsCount = dataTest('exercise-statistics-sets-count');
const repeatsCount = dataTest('exercise-statistics-repeats-count');
const setsDuration = dataTest('exercise-statistics-sets-duration');
const repeatsDuration = dataTest('exercise-statistics-repeats-duration');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(ExerciseStatistics, {
    props: {
      statistics: EXERCISES_STATISTICS_FIXTURE,
    },
  });
});

enableAutoUnmount(afterEach);

describe('ExerciseStatistics', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ExerciseStatistics)).toBeTruthy();
  });

  it('shows statistics rows', async () => {
    expect(wrapper.findAll(tableRow).length).toBe(EXERCISES_STATISTICS_FIXTURE.length);
  });

  it('shows statistics title', async () => {
    expect(wrapper.find(title).text()).toBe(EXERCISES_STATISTICS_FIXTURE[0].title);
  });

  it('shows sets and repeats count', async () => {
    expect(wrapper.find(setsCount).text()).toBe(EXERCISES_STATISTICS_FIXTURE[0].sets.toString());
    expect(wrapper.find(repeatsCount).text()).toBe(EXERCISES_STATISTICS_FIXTURE[0].repeats.toString());
  });

  it('shows sets and repeats average duration', async () => {
    expect(wrapper.find(setsDuration).text()).toBe(getAverageDuration(EXERCISES_STATISTICS_FIXTURE[0], 'set'));
    expect(wrapper.find(repeatsDuration).text()).toBe(getAverageDuration(EXERCISES_STATISTICS_FIXTURE[0], 'repeat'));
  });
});
