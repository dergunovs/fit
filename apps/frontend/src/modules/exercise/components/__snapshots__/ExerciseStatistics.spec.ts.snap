// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExerciseStatistics > matches snapshot 1`] = `
"<div>
  <table headers="[object Object],[object Object],[object Object]" lang="ru">
    <tbody>
      <tr data-test="exercise-statistics-table-row">
        <td><button layout="plain" iswrap="" data-test="exercise-statistics-title" class="_title_005ffb" data-equipment="true">Отжимание от скамьи</button></td>
        <td>
          <div class="_cell_005ffb">
            <div class="_count_005ffb" data-test="exercise-statistics-sets-count">35<dynamic-percent-stub percent="74"></dynamic-percent-stub>
            </div>
            <div class="_duration_005ffb" data-test="exercise-statistics-sets-duration">51.1с</div>
          </div>
        </td>
        <td>
          <div class="_cell_005ffb">
            <div class="_count_005ffb" data-test="exercise-statistics-repeats-count">578<dynamic-percent-stub percent="81"></dynamic-percent-stub>
            </div>
            <div class="_duration_005ffb" data-test="exercise-statistics-repeats-in-set">17</div>
          </div>
        </td>
      </tr>
      <tr data-test="exercise-statistics-table-row">
        <td><button layout="plain" iswrap="" data-test="exercise-statistics-title" class="_title_005ffb" data-equipment="true">Подтягивание</button></td>
        <td>
          <div class="_cell_005ffb">
            <div class="_count_005ffb" data-test="exercise-statistics-sets-count">30<dynamic-percent-stub percent="-10"></dynamic-percent-stub>
            </div>
            <div class="_duration_005ffb" data-test="exercise-statistics-sets-duration">40.9с</div>
          </div>
        </td>
        <td>
          <div class="_cell_005ffb">
            <div class="_count_005ffb" data-test="exercise-statistics-repeats-count">374<dynamic-percent-stub percent="-33"></dynamic-percent-stub>
            </div>
            <div class="_duration_005ffb" data-test="exercise-statistics-repeats-in-set">12</div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
  <div modelvalue="false" width="360" data-test="exercise-statistics-modal"></div>
</div>"
`;
